design/milestone_1_2_design_doc.md

```diff
@@ -0,0 +1,455 @@
+
+# Secrets Provider pod defaults
+secrets-provider:
+	image: cyberark/secrets-provider-for-k8s
+	name: cyberark-secrets-provider
+	namespace: default
+	ttlSecondsAfterFinished: 0	# Length at which Job should stay alive even after finishing task. Used for logging.
+	
+# Secrets Provider environment variable defaults
+secrets-provider-environment:
+	conjurApplianceUrl: https://conjur-follower.default.svc.cluster.local
+	conjurAuthnUrl: https://conjur-follower.default.svc.cluster.local/authn-k8s/authn-default
+	conjurAuthnLogin: host/conjur/authn-k8s/authn-default/default
+	conjurAccount: default
+	conjurSslCertificate: conjur-master-ca-env
+	debug=true
+	k8sSecrets=[]
+	containerMode=init
+	secretsDestination=k8s_secrets
+	rotationSupport=false
+```
+
+Customers can override these defaults by loading the Chart with `custom-values.yaml`. Helm will first look for values in the customer defined `custom-values.yaml`. If they do not exist, Helm will resort to the defaults configured in `values.yaml` and populate the manifest with those values. 
+
+For more detailed information on how a customer will install our Chart, see the section on [Installation](#installation).
+
+Variables from the `values.yaml` / `custom-values.yml `are applied for our project as follows:
+
+*templates/secrets-access-role.yaml*
+
+```
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ .Values.permission.serviceAccount }}
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ .Values.permission.roleName }}
+rules:
+  - apiGroups: [""]
+    resources: {{ .Values.permission.rules.resources }}
+    verbs:
+    	rules:
+			resources: secrets
+			verbs: [ "get", "update" ]
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  namespace: {{ .Values.secrets-provider.namespace }}
+  name: {{ .Values.permission.roleBinding }}
+subjects:
+  - kind: ServiceAccount
+    namespace: {{ .Values.secrets-provider.namespace }}
+    name: {{ .Values.permission.serviceAccountName }}
+roleRef:
+  kind: Role
+  apiGroup: rbac.authorization.k8s.io
+  name: {{ .Values.permission.roleName }}
+```
+
+*templates/secrets-provider.yaml*
+
+The following template takes into account the anticipated flows for Milestone 1 and 2.
+
+```yaml
+---
+{{- if eq .Values.secrets-provider-environment.rotationSupport "false" }}
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: {{ .Release.Name }}
+  namespace: {{ .Values.secrets-provider.namespace }}
+spec:
+	ttlSecondsAfterFinished: {{ .Values.secrets-provider.ttlSecondsAfterFinished }}
+{{ else }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ .Release.Name }}
+  namespace: {{ .Values.secrets-provider.namespace }}
+spec:
+{{- end }}
+  template:
+    serviceAccountName: {{ .Values.serviceAccount }}
+    containers:
+    - image: {{ .Values.secrets-provider.image }}
+        imagePullPolicy: Always
+        name: {{ .Values.secrets-provider.name }}
+        env:
```

Comment:
I think that the following are missing from the env:
- `SECRETS_DESTINATION` - is there a default value for this?
- `CONTAINER_MODE` (note that the default value is `sidecar`. what is the value that you want to have?)
- `K8S_SECRETS`
  - I know that `K8S_SECRETS_LABELS` replaces this. Do you think we should add this as an optional param?

Reply:
Good catch! Fixing this now.

That reminds me...
In order for a customer to add a new K8s Secret to the K8S_SECRETS, they will need to tear down the Chart and reinstall it with the updated list of `custom-values.yaml`. AKA tear down the deployment
Adding this to the limitations...

Reply:
> I know that `K8S_SECRETS_LABELS` replaces this. Do you think we should add this as an optional param?

I want to clarify, adding `K8S_SECRETS_LABELS` doesn't replace the ability for the customer to supply use with `K8S_SECRETS` values. If a customer would supply us with both, we would take both values and query for them in the K8s namespace.
I forgot `K8S_SECRETS` though, adding now!

Reply:
I am however wrap it in an if clause like the following

```
 {{- if eq .Values.secrets-provider-environment.k8sSecrets }}
   - name: K8S_SECRETS
     value: {{ .Values.secrets-provider-environment.k8sSecrets }}
{{- end }}			 
```

b/c in later milestones it is not required and it is enough that a customer will just have `K8S_SECRETS_LABELS`

Reply:
yes, i guess both need to be conditional.

Reply:
deleting this conditional for now because we are focusing on milestone 1 charts. This conditional is relevant for milestone 2

Reply:
@sigalsax can you please organise this thread and see all concerns have been addressed?
It got multiple subjects already

Reply:
>I want to clarify, adding K8S_SECRETS_LABELS doesn't replace the ability for the customer to supply use with K8S_SECRETS values

Shouldn't we deprecate K8S_Secrets? it seems we are adding complexity to our product. IMHO we should at least log a warning if the customer use it that it might be deprecated in the future

Reply:
We should decide with @InbalZilberman if we are going to support K8S_SECRETS. Currently there is misalignment in this area

Reply:
Our plan for the `K8S_SECRETS_LABELS` was not to deprecate the `K8S_SECRETS`, rather allow them to live side by side. We spoke about K8S_SECRETS_LABELS being an enhancement for a later milestone but will bring this up again